### DISCUSSION TITLE
how to use Tasmota Client with reading / setting parameters via mqtt

### DISCUSSION THREAD

**Discussion by bltDefender** [arendst/Tasmota#11493] [Q&A] [answered]
My goal is to have an D1 mini based tasmota master handling wifi/mqtt to pass through data to and from an arduino. That arduino controls a heat bed so it needs to receive the correct temperature. On a node-red dashboard I can set this temperature, it will be sent via mqtt to the tasmota master and it will pass through to the arduino. The arduino responses with the current temperature it measures from the ntc of the head bed. Tasmota master shall send back this temperature as is to the mqtt server so my dashboard can read it. 

steps I took: compile 9.2 with tasmota client enabled. client boud rate is 57600 so I made sure the correct values are set as well. On arduino I deployed the tasmota client lib (thank you Andre) and pushed the examples ino files successfully to the mcu (via direct FTDI line NOT via tasmota master firmware upload). So far I dont see any problems. Via mqtt I sent cmnd/myDeviceName/ClientSend <<content>> and on the console I see that this command has been successfully processed. 

I need help with the following steps:
1. Lets say, I send the command like cmnd/myDeviceName/ClientSend {"targetTemperature": 42}. How would I be able to receive this on the arduino so the loop can consider this as the new target temp?
2. I suppose I would use the tele command to send back the current temperature but so far I did not manage to receive this on the tasmota master. At least I would expect the console to display any received message coming in. Nothing there.
3. There is always being transmitted "something". On the arduino I simply check the income on the Serial connection 
`  if (Serial.available() > 0) {
    incomingByte = Serial.read();
    char myjson[100];
    sprintf(myjson, "{\"payload\":\" %u \"}", incomingByte);
    Serial.println(incomingByte, DEC); 
  }`
  That permanently shows 
`�{"payload":" 252 "}�252
�{"payload":" 252 "}�252
�{"payload":" 252 "}�252
�{"payload":" 252 "}�252
�{"payload":" 252 "}�252
�{"payload":" 252 "}�252
�{"payload":" 252 "}�252
�{"payload":" 252 "}�252
`
Always first byte only just to demonstrate that there is something. 
4. flashing via tasmota master does not work. I think as long as I do not see any working data exchange this is to put on second prio.
5. after having all wired (correctly) I cannot flash the arduino anymore. I do have to disable the D1 mini otherwise flashing directly from arduino ide fails. I think this because of 3. Apparently there is always some data being transmitted so data communication is broken by "attack" :)

As I see it, I am stuck here so any help is appreciated :)

**Comment by cn-superman** [comment, marked as answer]
#include <XPString.h>
#include <Arduino.h>

#include <TasmotaClient.h>
TasmotaClient client(&Serial);


void setup() {

  pinMode(A3, OUTPUT);
  pinMode(A0, INPUT);
  Serial.begin(57600);
  client.attach_FUNC_COMMAND_SEND(user_FUNC_RECEIVE);

}

void loop()  {

  client.loop();

}

void user_FUNC_RECEIVE(char *data)
{
  char epromData[40];
  XPString myStr(epromData, 40);

  myStr = data;

  if (myStr.starts_with("GET") == true || myStr.starts_with("get") == true)
  {
    myStr.remove(0, 4);
    int tmpDly = atoi(myStr);

    delay(10);

    int xxx = 178;

    myStr = "{\"temp\":\"";
    myStr += xxx ;
    myStr += "\"}";

    client.SendTele(myStr);

  }

  else
  {
    myStr = "\"Error command.\"";
    client.SendTele(myStr);

  }

}



The sample code above uses the library: xpstring, please download it from the library manager automatically.
have to be aware of is:
1. Do not use the hardware serial port of 8266MCU, you will encounter the problem of garbled characters (meaning that you cannot use modules like 826601 or 826601S)
2. Use the sendtel command to publish to the MQTT server itself as a JSON format string. Therefore, when generating the reply string, prepare the JSON format of the second-level key and value.
3. Once the program has been flashed using arduino IDE (BOOTLOADER is already included), the firmware of ARDUINO can be flashed through TASMOTA's OTA in the future.
Hope to help you~~

**Reply by bltDefender** [reply]
awesome. Thank you very much, will implement like you suggested.

**Reply by bltDefender** [reply]
that works thank you. One thing open: I cannot do the OTA via tasmota. it presents error 301 after some seconds. I see the reset and then it fails. Weird thing is, that suddenly ClientSend does not word instead, SerialSend works on the Client PINs as well. baudrate again is 115200. checked the wiring, its okay. Will try a brand new d1 mini to exclude faulty hardware as a reason.

**Reply by bltDefender** [reply]
fresh installation, new chip. Client now works as expected but there is permanent data output on the defined TX PIN. I checked with platformio monitor, it looks like this:
![image](https://user-images.githubusercontent.com/37219611/112762800-06011280-9002-11eb-9033-6cdafee63c24.png)
It is not the native Serial Pins, but D1 and D2 in this case. Does this belong to the Client and would be correct? Or is this wrong?

**Reply by bltDefender** [reply]
you see this 45? This where I did ClientSend 45 in the console
